std-bits/database/src/main/java/org/enso/database/audit/AuditedConnection.java

```diff
@@ -0,0 +1,369 @@
+package org.enso.database.audit;
+
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.CallableStatement;
+import java.sql.Clob;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.NClob;
+import java.sql.PreparedStatement;
+import java.sql.SQLClientInfoException;
+import java.sql.SQLException;
+import java.sql.SQLWarning;
+import java.sql.SQLXML;
+import java.sql.Savepoint;
+import java.sql.Statement;
+import java.sql.Struct;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Executor;
+
+abstract class AuditedConnection implements Connection {
+  protected final Connection underlying;
+
+  AuditedConnection(Connection underlying) {
+    this.underlying = underlying;
+  }
+
+  abstract void auditQuery(String operationType, String sql);
+
+  abstract void auditTransaction(String operation);
+
+  private RuntimeException unimplemented(String name) {
+    throw new UnsupportedOperationException(
+        name + " is not implemented. This is a bug in the Database library.");
+  }
+
+  @Override
+  public Statement createStatement() throws SQLException {
+    return new AuditedStatementImpl(underlying.createStatement());
+  }
+
+  @Override
+  public PreparedStatement prepareStatement(String sql) throws SQLException {
+    return new AuditedPreparedStatementImpl(underlying.prepareStatement(sql), sql);
+  }
+
+  @Override
+  public CallableStatement prepareCall(String sql) throws SQLException {
+    throw unimplemented("prepareCall");
+  }
+
+  @Override
+  public String nativeSQL(String sql) throws SQLException {
+    return underlying.nativeSQL(sql);
+  }
+
+  @Override
+  public void setAutoCommit(boolean autoCommit) throws SQLException {
+    if (autoCommit != underlying.getAutoCommit()) {
+      auditTransaction("setAutoCommit " + autoCommit);
+    }
+    underlying.setAutoCommit(autoCommit);
+  }
+
+  @Override
+  public boolean getAutoCommit() throws SQLException {
+    return underlying.getAutoCommit();
+  }
+
+  @Override
+  public void commit() throws SQLException {
+    auditTransaction("commit");
+    underlying.commit();
+  }
+
+  @Override
+  public void rollback() throws SQLException {
+    auditTransaction("rollback");
+    underlying.rollback();
+  }
+
+  @Override
+  public void close() throws SQLException {
+    underlying.close();
+  }
+
+  @Override
+  public boolean isClosed() throws SQLException {
+    return underlying.isClosed();
+  }
+
+  @Override
+  public DatabaseMetaData getMetaData() throws SQLException {
+    return underlying.getMetaData();
+  }
```

Comment:
Currently metadata is not audited. That means that the act of listing tables existing in the database is _not_ audited.

Do you think we should change that? It was unclear to me if it's something that we want to log or not. It is easy to add if we decided that we _do_ want it.

Reply:
Agree don't think we need to audit Metadata queries.